### DISCUSSION TITLE
custom ui

### DISCUSSION THREAD

**Discussion by Dairoe** [vlang/v#16979] [Questions and Answers]
honestly lang is great , but how would one go about creating custom graphic lib, the one in the vpm is not great it need some work, which is why i want try at creating custom one , unless i can do all the custom drawing on the vpm ui lib,

idk c so would converting a lib to v even help? I rather not rely on the c bindings but v bindings

**Comment by JalonSolov** [comment]
What, exactly, do you mean by "graphic lib"?  Just drawing general shapes?  If so, the `gg` module that comes with V does a decent job with this.  Look for things in the `examples` directory.

If you mean a lib to do UI work, there are at least 3 I know of:

https://github.com/vlang/ui
https://github.com/malisipi/mui
https://github.com/IsaiahPatton/ui

**Reply by Dairoe** [reply]
basically do own custom drawing like skia for buttons etc..

 the vlang/ui i aready messed around with it doesn't work well out the box atleast on linux, though i haven't dived deep in the gg module i take another look at nd also try out the IsaiahPatton/ui

**Reply by Dairoe** [reply]
follow up patton lib seems like a port with custom themes from vlang/ui which doesnt have any new changes except the way its called

**Reply by Wajinn** [reply]
> basically do own custom drawing like skia for buttons etc..

That's what gg and gx are for.  If a person takes the time to look at V examples, it will become more clear.  A good example is [2048](https://github.com/vlang/v/tree/master/examples/2048).  In exploring this and some other examples, it can be seen how to make something customized to one's liking.

Also [MUI](https://github.com/malisipi/mui) was mentioned too, and it seems it was overlooked or unexplored.  However, it has custom widget functionality and is designed differently than [Isaiah's UI](https://github.com/IsaiahPatton/ui).

**Reply by Dairoe** [reply]
yes ive seen all the examples of ui even the 2048 also toyed with every single example even checked the source codes , things these so called ui lib in v have in common is absurd they are not different, a simple ui should have scalable things outside the box, etc.. 

the sokol lib is what im kinda looking for so im testing it , though it riddled with c calls im better off using c ,

**Reply by JalonSolov** [reply]
gg uses sokol for all the rendering already.  Both are available in the V stdlib.

**Reply by Wajinn** [reply]
> yes ive seen all the examples...
> 
> ...the sokol lib is what im kinda looking for so im testing it, though it riddled with c calls im better off using c

Sokol (as a third party tool) is used by various languages, so the situation would be similar, or possibly even less convenient.  V has additional libraries to create greater ease of use.  Something purely V would be great, but in order to have such cross-platform graphical functionality at the lower levels will be very hard, need more manpower, be arguably redundant, and/or is arguably unnecessary to avoid C.  

The answers given are with the goal of being helpful, and to show what is presently available for V.  In my personal research, I found V's gg, gx, and various tools to be at least useful and convenient (relative to the age of V as a developing language).  But, clearly everyone has different standards.  Every preference or everyone's satisfaction can't be guaranteed.
